doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,414 @@
+objects should subclass another class, such as {class}`~.Boolean`,
+{class}`~.MatrixExpr`, {class}`~.Expr`, or {class}`~.Basic`. Much of what is
+written here only applies to {class}`~.Function` subclasses and will not work
+for general {class}`~.Basic` or {class}`~.Expr` subclasses.
+
+## Easy Cases: Fully Symbolic or Fully Evaluated
+
+Before digging into the more advanced functionality for custom functions, we
+should mention two common cases that are much simpler, the case where the
+function is fully symbolic, and the case where the function is fully
+evaluated.
+
+(custom-functions-fully-symbolic)=
+### Fully Symbolic
+
+If your function `f` has no mathematical properties you
+want to define on it, and should never evaluate on any arguments, you can
+create an undefined function using `Function('f')`
+
+```
+>>> from sympy import symbols, Function
+>>> x = symbols('x')
+>>> f = Function('f')
+>>> f(x)
+f(x)
+>>> f(0)
+f(0)
+```
+
+This is useful, for instance, when solving [ODEs](ode-docs).
+
+This is also useful if you only wish to create a symbol that depends on
+another symbol for the purposes of differentiation. By default, SymPy assumes
+all symbols are independent of one another:
+
+```
+>>> from sympy.abc import x, y
+>>> y.diff(x)
+0
+```
+
+To make a symbol depend on another symbol, you should make it a function that
+explicitly depends on that symbol.
+
+```
+>>> y = Function('y')
+>>> y(x).diff(x)
+Derivative(y(x), x)
+```
+
+If you want your function to have additional behavior, for example, to have a
+custom derivative, or to evaluate on certain arguments, you should create a
+custom `Function` subclass as described below. However, undefined functions do
+support one additional feature, which is that assumptions can be defined on
+them, using the same syntax as used by symbols. This defines the assumptions
+of the output of the function, not the input (that is, it defines the
+function's range, not its domain).
+
+```
+>>> g = Function('g', real=True)
+>>> g(x)
+g(x)
+>>> g(x).is_real
+True
+```
+
+To make a function's assumptions depend on its input in some way, you should
+create a custom `Function` subclass as described below.
+
+(custom-functions-fully-evaluated)=
+### Fully Evaluated
+
+At the other end of the spectrum are functions that always evaluate to
+something no matter what their inputs are. These functions are never left in
+an unevaluated, symbolic form like `f(x)`.
+
+In this case, you should just use a normal Python function using the `def`
+keyword.
+
+```
+>>> def f(x):
+...     if x == 0:
+...         return 0
+...     else:
+...         return x + 1
+>>> f(0)
+0
+>>> f(1)
+2
+>>> f(x)
+x + 1
+```
+
+If you find yourself defining an `eval` method on a `Function` subclass where
+you always return a value and never return `None`, you should consider just
+using a normal Python function instead, as there is no benefit to using a
+symbolic `Function` subclass in that case.
```

Comment:
There is potential benefit if you can use `evaluate=False` and also have printing support. Another point to consider is that maybe it should be possible to call the function with symbols as arguments so the eval method should maybe handle that.

Reply:
But do we really want to recommend that? evaluate=False is very fragile. It's a much better design to just avoid automatic evaluation and move it to some other method such as `doit`. Even for something like printing, it can be more complicated than it might seem, because that doesn't really tell the whole story. What if you want to manipulate the expression somewhat to get it in some form to print. What if you later decide to do some stuff that isn't printing. The more code you have build up around a function, the harder it is to remove automatic evaluation from it. That's my experience with SymPy functions and I believe it is also true of user defined functions.

And even if you are just doing printing and nothing else wouldn't it still be simpler to just have a function that doesn't evaluate (in fact, you could just use a `Function('f')` like I described earlier in the guide) than passing evaluate=False everywhere?

> Another point to consider is that maybe it should be possible to call the function with symbols as arguments so the eval method should maybe handle that.

I'm not following this. Both a `def` function and an `eval` method of a `Function` subclass should handle symbols as arguments.

Reply:
The only benefit I can see is maybe if we implement https://github.com/sympy/sympy/issues/4787, you could have a symbolic function object `f` which always evaluates. But even then, I don't know if having functions as objects that can be inside of other expressions is something that too many people really need. https://github.com/sympy/sympy/issues/4787 is important, but that's the most important part of it. Not many people request that feature.

The reason I am stressing the "don't just return an expression in `eval`" and also the "use a `def` function if you just want a simple shortcut" is because it's a common thing I see people do on StackOverflow and the mailing list. I think part of it is that people don't really sit down and think about what they actually want, but also part of it is that they don't completely understand how `eval` works and maybe don't even realize that just returning an expression in `eval` has this implication. 

Finally, I would hope that everything in this guide should apply equally well to functions in library code as to user defined functions, and indeed I expect it to be useful for contributors. So I'm trying to apply the highest standard to the recommendations here. Users are of course free to do whatever they want, but we should not recommend something that we wouldn't do ourselves. The goal is to replace this wiki page, which isn't written very well https://github.com/sympy/sympy/wiki/Development-Tips-:-About-implementing-special-functions.

Reply:
I agree that automatic evaluation should be discouraged for symbolic functions. My suggestion though is that if someone is in the position of writing a Function subclass with an eval method that always returns maybe they should consider changing it so that the eval method does not always return so that then their function can also work with symbolic arguments.

Reply:
I think which one you really want requires introspection as the whether you're looking for a symbolic function or just a shorthand. I can try to make it clearer here that there are two options.